model/scope.go

```diff
@@ -0,0 +1,457 @@
+	ScopeAnyResource  ScopeResource  = "*"
+	ScopeSeparator                   = ":"
+)
+
+const (
+	ScopeChannels ScopeResource = "channels"
+	ScopeCommands ScopeResource = "commands"
+	ScopeEmojis   ScopeResource = "emojis"
+	ScopeFiles    ScopeResource = "files"
+	ScopeOAuth2   ScopeResource = "oauth2"
+	ScopePosts    ScopeResource = "posts"
+	ScopeTeams    ScopeResource = "teams"
+	ScopeUsers    ScopeResource = "users"
+	ScopePlugins  ScopeResource = "plugins"
+)
+
+const (
+	ScopeCreate  ScopeOperation = "create"
+	ScopeDelete  ScopeOperation = "delete"
+	ScopeExecute ScopeOperation = "execute"
+	ScopeJoin    ScopeOperation = "join"
+	ScopeManage  ScopeOperation = "manage"
+	ScopeRead    ScopeOperation = "read"
+	ScopeSearch  ScopeOperation = "search"
+	ScopeUpdate  ScopeOperation = "update"
+)
+
+const (
+	ScopeChannelsAny    Scope = "channels:*"
+	ScopeChannelsCreate Scope = "channels:create"
+	ScopeChannelsDelete Scope = "channels:delete"
+	ScopeChannelsRead   Scope = "channels:read"
+	ScopeChannelsSearch Scope = "channels:search"
+	ScopeChannelsUpdate Scope = "channels:update"
+	ScopeChannelsJoin   Scope = "channels:join"
+
+	ScopeCommandsExecute Scope = "commands:execute"
+
+	ScopeEmojisAny    Scope = "emojis:*"
+	ScopeEmojisCreate Scope = "emojis:create"
+	ScopeEmojisDelete Scope = "emojis:delete"
+	ScopeEmojisRead   Scope = "emojis:read"
+	ScopeEmojisSearch Scope = "emojis:search"
+	ScopeEmojisUpdate Scope = "emojis:update"
+
+	ScopeFilesAny    Scope = "files:*"
+	ScopeFilesCreate Scope = "files:create"
+	ScopeFilesDelete Scope = "files:delete"
+	ScopeFilesRead   Scope = "files:read"
+	ScopeFilesSearch Scope = "files:search"
+	ScopeFilesUpdate Scope = "files:update"
+
+	ScopeOAuth2Manage Scope = "oauth2:manage"
+
+	ScopePostsAny    Scope = "posts:*"
+	ScopePostsCreate Scope = "posts:create"
+	ScopePostsDelete Scope = "posts:delete"
+	ScopePostsRead   Scope = "posts:read"
+	ScopePostsSearch Scope = "posts:search"
+	ScopePostsUpdate Scope = "posts:update"
+
+	ScopeTeamsAny    Scope = "teams:*"
+	ScopeTeamsCreate Scope = "teams:create"
+	ScopeTeamsDelete Scope = "teams:delete"
+	ScopeTeamsRead   Scope = "teams:read"
+	ScopeTeamsSearch Scope = "teams:search"
+	ScopeTeamsUpdate Scope = "teams:update"
+	ScopeTeamsJoin   Scope = "teams:join"
+
+	ScopeUsersAny    Scope = "users:*"
+	ScopeUsersCreate Scope = "users:create"
+	ScopeUsersDelete Scope = "users:delete"
+	ScopeUsersRead   Scope = "users:read"
+	ScopeUsersSearch Scope = "users:search"
+	ScopeUsersUpdate Scope = "users:update"
+)
+
+var validScopes = map[ScopeResource][]ScopeOperation{
+	ScopeChannels: append(crudsScopeOps, ScopeJoin),
+	ScopeCommands: {ScopeExecute},
+	ScopeEmojis:   crudsScopeOps,
+	ScopeFiles:    crudsScopeOps,
+	ScopeOAuth2:   {ScopeManage},
+	ScopePosts:    crudsScopeOps,
+	ScopeTeams:    append(crudsScopeOps, ScopeJoin),
+	ScopeUsers:    crudsScopeOps,
+}
+
+var crudsScopeOps = []ScopeOperation{ScopeCreate, ScopeRead, ScopeUpdate, ScopeDelete, ScopeSearch}
```

Comment:
It confused me that `crudsScopeOps` is defined below where it's used here

Reply:
done